dbos-rules.test.ts

```diff
@@ -0,0 +1,220 @@
+import { dbosRulesPerName } from "./dbos-rules";
+
+RuleTester.it = vitest.it;
+RuleTester.itOnly = vitest.it.only;
+RuleTester.describe = vitest.describe;
+RuleTester.afterAll = vitest.afterAll;
+
+//////////
+
+// https://stackoverflow.com/questions/51851677/how-to-get-argument-types-from-function-in-typescript
+type ArgumentTypes<F extends Function> = F extends (...args: infer A) => any ? A : never;
+
+// https://stackoverflow.com/questions/41253310/typescript-retrieve-element-type-information-from-array-type
+type ArrayElementType<ArrayType extends readonly unknown[]> = ArrayType extends readonly (infer T)[] ? T : never;
+
+type TestTypes = ArgumentTypes<typeof tester.run>[2];
+type SuccessTests = TestTypes["valid"];
+type FailureTests = TestTypes["invalid"];
+
+type TestSet = [string, SuccessTests, FailureTests][];
+type SuccessTest = ArrayElementType<SuccessTests>;
+type FailureTest = ArrayElementType<FailureTests>;
+
+function doTest(title: string, successTests: SuccessTests, failureTests: FailureTests) {
+  const ruleName = "detect-nondeterministic-calls";
+  tester.run(title, dbosRulesPerName[ruleName], { valid: successTests, invalid: failureTests });
+}
+
+//////////
+
+const tester = new RuleTester({
+  parser: "@typescript-eslint/parser",
+  parserOptions: { project: "tsconfig.json" },
+  defaultFilenames: { ts: "dbos-rules.test.ts", tsx: "this_file_doesnt_exist.tsx" }
+});
+
+////////// These functions build different types of test cases with some primitive code structure around them.
+
+function makeExpectedDetCode(
+  code: string,
+  codeAboveClass: string,
+  enclosingFunctionParams: string): string {
+
+  return `
+    ${codeAboveClass}
+
+    class Foo {
+      @Workflow()
+      foo(${enclosingFunctionParams}) {
+        ${code}
+      }
+    }
+  `;
+}
+
+function makeExpectedSuccessTest(code: string,
+  { codeAboveClass, enclosingFunctionParams } = { codeAboveClass: "", enclosingFunctionParams: "" }): SuccessTest {
+
+  return { code: makeExpectedDetCode(code, codeAboveClass, enclosingFunctionParams) };
+}
+
+function makeExpectedFailureTest(code: string, expectedErrorIds: string[],
+  { codeAboveClass, enclosingFunctionParams } = { codeAboveClass: "", enclosingFunctionParams: "" }): FailureTest {
+
+  const inObjectFormat = expectedErrorIds.map((id) => { return { messageId: id }; });
+  return { code: makeExpectedDetCode(code, codeAboveClass, enclosingFunctionParams), errors: inObjectFormat };
+}
+
+//////////
+
+const testSet: TestSet = [
+  ["global mutations", [],
+
+    [makeExpectedFailureTest(
+      `
+      let x = 3;
+      let y = {a: 1, b: 2};
+
+      class Bar {
+        @Workflow()
+        foo() {
+          x = 4; // Not allowed
+          y.a += 1; // Not allowed
+
+          let y = {a: 3, b: 4}; // Aliases the global 'y'
+          y.a = 1; // Not a global modification anymore
+        }
+
+        bar() {
+          y.b += 2;
+          let z = 8;
+
+          class Bar {
+            @Workflow()
```

Comment:
I am not sure we should even let them do this, but probably nobody will try.

Reply:
Yes - just making sure that mutation detection of classes in functions in classes (etc.) works